### PR TITLE
CRM-19630 optionally allow bi-directional relation

### PR DESCRIPTION
**Pull request by jitendrapurohit**
rebased and updated https://github.com/civicrm/civicrm-core/issues/9369 with the following fixes.

1) Fatal error on disable of relationship type.
2) Optimization and removal of unwanted codes.
3) Added unit test.

* [CRM-19630: Optionally allow bi-directional relationships](https://issues.civicrm.org/jira/browse/CRM-19630)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I feel like you should solicit some feedback  - on dev list or at least mm chanel about this. I'm not sure if there would be concerns or what they might be but it feels like the sort of thing people should be able to input on

**Comment by seanmadsen** [comment]
I support this change conceptually, as Pete described it [on Mattermost](https://chat.civicrm.org/civicrm/pl/8zysh45uuiyymdawgjnakuo88h). (I have not tested it though.)

**Request:** Can you update the user documentation along with this change? Here is the relevant section: 
https://docs.civicrm.org/user/en/latest/organising-your-data/relationships/#creating-new-relationship-types

I think it would be nice to include Pete's example of a "favorite" type. It's a great example that helps me easily understand why this checkbox would exist.

If you need help editing these docs, I can help. Here is our [meta-documentation about writing documentation](https://docs.civicrm.org/dev/en/latest/documentation/). 

Basically what I'd like to see here is a PR made [here](https://github.com/civicrm/civicrm-user-guide/pulls) which references *this* PR.

**Review by xurizaemon — COMMENTED** [review]
added a couple of comments, based on reading not running the code

I wonder about new things that come into scope if Relationship is extended to new things.

* Should relations pointing to a contact be visible on that contact record? Currently the UI does this, but for some uses that wouldn't be appropriate or desirable (for a much-⭐ed  contact, we'd hit resource issues loading their many ⭐ers; for a contact that has been flagged for abuse by the same means, we might not want to reveal flaggers). These use cases might not be essential to address in this PR, since that's more about how people will use the feature.
* Can a contact relate to themself with this new definition or is that excluded? How does CiviCRM behave when asked to relate contact A to contact A?
* What existing design is there for how Relationship works, and what is the new design here?

(Does _relationship_ have a bidirectional connotation, where a _relation_ doesn't? I think the term "relation" describes (say) me marking a contact as an influence, or a favourite, or flagging their record for some reason. But I wouldn't feel like being a fan of a famous person counts as a "relationship", since it's one-sided.)

File: templates/CRM/Admin/Form/RelationshipType.tpl [hunk 1]
```diff
@@ -61,6 +61,12 @@
                 <td class="label">{$form.is_active.label}</td>
                 <td>{$form.is_active.html}</td>
             </tr>
+            <tr class="crm-relationship-type-form-block-is_bidirectional">
+                <td class="label">{$form.is_bidirectional.label}</td>
+                <td>{$form.is_bidirectional.html}<br />
+                  <span class="description">{ts}Can a 'B to A' relationship be created when their is's an existing 'A to B' relationship?{/ts}</span>
```
Comment (jitendrapurohit) [inline templates/CRM/Admin/Form/RelationshipType.tpl]: Message updated. Thanks!

File: tests/phpunit/api/v3/RelationshipTest.php [hunk 2]
```diff
@@ -1292,4 +1292,38 @@ public function testCreateRelatedMembership() {
...
+    $params = array(
+      'contact_id_a' => $this->_cId_a_2,
+      'contact_id_b' => $this->_cId_a,
+      'relationship_type_id' => $relType,
+    );
+
+    $this->callAPISuccess('Relationship', 'create', $params);
+    $this->relationshipTypeDelete($relType);
+  }
+
```
Comment (jitendrapurohit) [inline tests/phpunit/api/v3/RelationshipTest.php]: Extended `testRelationshipCreateAlreadyExists()` to include the above check.

**Review by JohnFF — CHANGES_REQUESTED** [review]
Hi,

I think this is a potentially useful feature. Without having tested it, I'd make the following comments:

1. There should also be a Unit Test for relationships where the contact types are different.
2. There shouldn't be any references to CRM-19630. The comment should say what the block of code does if it's unclear. If there's a long block of code then you can put it in a separate function, or put // Start handling  bidirectional relationships around // End handling bidirectional relationships. Nothing here is really long enough to justify its own function.

**Comment by eileenmcnaughton** [comment]
" Nothing here is really long enough to justify its own function."

That's fighting talk that is! :-)

 I think just about anythin justifies it;s own function e.g

It's more readable to have 

``
$d=  $this->isASeparateTransaction() 
``
vs
``
// blah
// blah
// And so, in this weird & wonderful way we decide if it is a separate transaction
// due to the interpolation between leaves & smelly pits.
$d = $this->isMyMothersLeftArmpit < 25 && 9 %6 || $isTwoTreesHaveLeaves;
``

so even a single line that is really weird can justify it's own function for readability....

**Comment by JohnFF** [comment]
@eileenmcnaughton Very true, I misspoke. I really meant "requires" or "mandates". But yes, abstracting into functions == win. In this instance I still don't personally feel it's mandated or required. I'll give more feedback once I've given this PR a kicking IRL.

**Comment by jitendrapurohit** [comment]
migrated the upgrade part to 4.7.22.

**Comment by petednz** [comment]
@jitendrapurohit do you need to come back to this?

**Comment by eileenmcnaughton** [comment]
This has been around a long time & I'm not really sure what to do with it - I feel like it needs someone like @colemanw to confirm it is a change that should be reviewed & accepted for core. I'm going to close it for now because it's stale - conversation can still proceed on a closed PR & it's tracked through JIRA. I believe it will get more visibility if you re-open
